core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/PathTrackingInputFormat.java

```diff
@@ -125,125 +121,88 @@ public NullWritable getCurrentKey() throws IOException, InterruptedException {
-    private final RecordReader<AvroKey<GenericRecord>, NullWritable> delegate;
-    private final Schema schema;
-    private final String pathField;
-    private final String path;
+  private static class TrackingTextRecordReader extends TrackingRecordReader<LongWritable, Text> {
 
-    private TrackingAvroRecordReader(RecordReader<AvroKey<GenericRecord>, NullWritable> delegate,
-                                     @Nullable String pathField, String path) {
-      this.delegate = delegate;
-      this.pathField = pathField;
-      this.path = path;
-      schema = getOutputSchema(pathField);
+    private TrackingTextRecordReader(RecordReader<LongWritable, Text> delegate, @Nullable String pathField,
+                                     String path) {
+      super(delegate, pathField, path);
     }
 
-    @Override
-    public void initialize(InputSplit split, TaskAttemptContext context) throws IOException, InterruptedException {
-      delegate.initialize(split, context);
-    }
+    public StructuredRecord getCurrentValue(StructuredRecord.Builder recordBuilder) throws IOException, InterruptedException {
+      LongWritable key = delegate.getCurrentKey();
+      Text text = delegate.getCurrentValue();
 
-    @Override
-    public NullWritable getCurrentKey() throws IOException, InterruptedException {
-      return NullWritable.get();
+      recordBuilder.set("offset", key.get());
+      recordBuilder.set("body", text.toString());
+      return recordBuilder.build();
     }
+  }
 
-    @Override
-    public StructuredRecord getCurrentValue() throws IOException, InterruptedException {
-      AvroToStructuredTransformer recordTransformer = new AvroToStructuredTransformer();
-      return recordTransformer.transform(delegate.getCurrentKey().datum());
-    }
+  private static class TrackingAvroRecordReader extends TrackingRecordReader<AvroKey<GenericRecord>, NullWritable> {
+    private static AvroToStructuredTransformer recordTransformer;
 
-    @Override
-    public boolean nextKeyValue() throws IOException, InterruptedException {
-      return delegate.nextKeyValue();
+    private TrackingAvroRecordReader(RecordReader<AvroKey<GenericRecord>, NullWritable> delegate,
+                                 @Nullable String pathField, String path) {
+      super(delegate, pathField, path);
     }
 
     @Override
-    public float getProgress() throws IOException, InterruptedException {
-      return delegate.getProgress();
+    public void initialize(InputSplit split, TaskAttemptContext context) throws IOException, InterruptedException {
+      super.initialize(split, context);
+      recordTransformer = new AvroToStructuredTransformer();
     }
 
-    @Override
-    public void close() throws IOException {
-      delegate.close();
+    public StructuredRecord getCurrentValue(StructuredRecord.Builder recordBuilder) throws IOException, InterruptedException {
+      return recordTransformer.transform(delegate.getCurrentKey().datum());
```

Comment:
this also made me realize the getOutputSchema method in this class has not been changed, which means the schema in the builder is only correct for the 'text' format. 

Now that we support avro and parquet, there needs to be an additional 'schema' property added to the plugin's config. The schema minus the path field should be the schema used to read the data. 

For AvroKeyInputFormat, this means setting the 'avro.schema.input.key'. For AvroParquetInputFormat, this means setting the 'parquet.avro.read.schema'. You can see where those come from by reading the respective input format. If the source is not given a schema, it should use the current behavior, which uses the write schema as the read schema. 

You can handle the case where schema is not specified in a separate PR, as it involves modifying schemas on the fly, since schema is not longer constant.